### DISCUSSION TITLE
Scripting support - How to use it!

### DISCUSSION THREAD

**Discussion by JPersson77** [JPersson77/LGTVCompanion#135] [General]
### What is it?
LGTV Companion v2.2.0 and later can provide access via an API for asynchronously sending and receiving messages and events, via named pipes. This essentially allows external scripts and applications to act as clients and interact with LGTV Companion, and devices managed by it. 

The feature is currently to be considered in BETA and under development. Further development will be based on input from users on the implementation and its usability.

### What can it do?
V2.2.0 supports the following:
- READING global power and system events, such as screen on/off, user idle, suspend, resume, reboot, shutdown. 
- WRITING messages, currently consisting of the full set of command line parameters. 

### What can it not do?
Currently there are some important delimitations:
- Arbitrary control messages/requests cannot yet be sent to LG devices
- It is not possible yet to query LG devices, f e to get current power state, volume level, channel etc

### How is it implemented?
It is implemented using asynchronous named pipes, which means that most scripting- and programming languages can access it.

### What about examples on how to use?
There are two basic example scripts which show the current implementation here: [Example Scripts](https://github.com/JPersson77/LGTVCompanion/tree/master/Example%20scripts) 

The sample scripts implement a simple named pipes client using AutoHotKey (AHK) v2. 
- The first will show how to READ all the events which can currently be sent by LGTV Companion, i.e implement a basic loop for waiting for events and processing them. 
- The second is an example of how to WRITE messages to LGTV Companion, by setting up two keyboard shortcuts for muting (SHIFT+CTRL+M) and unmuting (SHIFT+CTRL+U) the built-in speakers of the TV. 
	
Like previously mentioned a named pipe client can be implemented in your favourite scripting-, or programming language, but if you want to try the examples ensure to install AutoHotKey by downloading the installer from here: https://www.autohotkey.com/ and then install it. (_AutoHotkey is a free, open-source scripting language for Windows that allows users to easily create small to complex scripts for all kinds of tasks such as: form fillers, auto-clicking, macros, etc.)_

	
**Remember to enable the "External API" option in LGTV Companion.**

**Comment by anzial** [comment]
Cool. I hope I'll be able to learn how to use with command line :)

**Reply by JPersson77** [reply]
Hi Anzial, you can use lgtv companion's commandline for posting the same messages to lgtv companion as you can via named pipes. This is probably 95% of the use case, but to also be able to receive events and/or data, named pipes is the only option at the moment. 

Or maybe you meant you want to use it in batch files, and not autohotkey? Then i am looking forward to hearing your progress. I just picked autohotkey for the examples since I find it easy to work with.

The support for messages you can send to the TV will be increased in the next version

**Reply by anzial** [reply]
Thanks for that! I'm actually using bat files to control on/off state of the TV. BTW, are there any more command line arguments which are yet to be listed in the description? :) As always, I'm interested in the ones controlling power as well as possibly picture modes

**Comment by JPersson77** [comment]
Thanks too! The description says it all for v2.2.2 but if you ping me on discord, https://discord.gg/7KkTPrP3fq , I can send you a development version which does a few more, including picture mode. Another user is currently testing it for creating a script for automatically setting a specific picture mode when a game is running.

**Reply by anzial** [reply]
hmm, automatic script sounds fantastic. I do actually switch between different picture modes between desktop usage and games, and media, using the remote for now but using your awesome app would make it even more awesome :)
